token/token.go

```diff
@@ -6,156 +6,157 @@ import (
+const (
+	// Service Account Names
+
+	Auth         = "fabric8-auth"
+	WIT          = "fabric8-wit"
+	OsoProxy     = "fabric8-oso-proxy"
+	Tenant       = "fabric8-tenant"
+	Notification = "fabric8-notification"
+	JenkinsIdler = "fabric8-jenkins-idler"
+	JenkinsProxy = "fabric8-jenkins-proxy"
+
+	devModeKeyID = "test-key"
+)
+
+// configuration represents configuration needed to construct a token manager
+type configuration interface {
 	GetAuthServiceURL() string
-	GetKeysTokenPath() string
-	DeveloperModeEnabled() bool
+	GetAuthKeysPath() string
+	GetDevModePrivateKey() []byte
 }
 
 // TokenClaims represents access token claims
 type TokenClaims struct {
-	Name          string                `json:"name"`
-	Username      string                `json:"preferred_username"`
-	GivenName     string                `json:"given_name"`
-	FamilyName    string                `json:"family_name"`
-	Email         string                `json:"email"`
-	Company       string                `json:"company"`
-	SessionState  string                `json:"session_state"`
-	Authorization *AuthorizationPayload `json:"authorization"`
+	Name          string `json:"name"`
+	Username      string `json:"preferred_username"`
+	GivenName     string `json:"given_name"`
+	FamilyName    string `json:"family_name"`
+	Email         string `json:"email"`
+	EmailVerified bool   `json:"email_verified"`
+	Company       string `json:"company"`
 	jwt.StandardClaims
 }
 
-// AuthorizationPayload represents an authz payload in the rpt token
-type AuthorizationPayload struct {
-	Permissions []Permissions `json:"permissions"`
-}
-
-// Permissions represents a "permissions" in the AuthorizationPayload
-type Permissions struct {
-	ResourceSetName *string `json:"resource_set_name"`
-	ResourceSetID   *string `json:"resource_set_id"`
-}
-
-type PublicKey struct {
-	KeyID string
-	Key   *rsa.PublicKey
+// Parser parses a token and exposes the public keys for the Goa JWT middleware.
+type Parser interface {
+	Parse(ctx context.Context, tokenString string) (*jwt.Token, error)
+	PublicKeys() []*rsa.PublicKey
 }
 
 // Manager generate and find auth token information
 type Manager interface {
+	Parser
 	Locate(ctx context.Context) (uuid.UUID, error)
 	ParseToken(ctx context.Context, tokenString string) (*TokenClaims, error)
-	PublicKey(kid string) *rsa.PublicKey
-	PublicKeys() []*rsa.PublicKey
-	IsServiceAccount(ctx context.Context, serviceName string) bool
+	ParseTokenWithMapClaims(ctx context.Context, tokenString string) (jwt.MapClaims, error)
+	PublicKey(keyID string) *rsa.PublicKey
+	AddLoginRequiredHeader(rw http.ResponseWriter)
 }
 
 type tokenManager struct {
 	publicKeysMap map[string]*rsa.PublicKey
-	publicKeys    []*PublicKey
+	publicKeys    []*jwk.PublicKey
+	config        configuration
 }
 
 // NewManager returns a new token Manager for handling tokens
-func NewManager(config tokenManagerConfiguration) (Manager, error) {
+func NewManager(config configuration) (Manager, error) {
```

Comment:
Also IMO it's worth adding DefaultManager we introduced recently in Auth: https://github.com/fabric8-services/fabric8-auth/blob/master/token/token.go#L55-L67

We needed it because sometimes we don't have a request context (generated by goa middleware with injected token manager) to extract the token manager from it - https://github.com/fabric8-services/fabric8-auth/blob/master/token/tokencontext/token_context.go#L21
For example when we refresh cluster cache in a goroutine every N minute. We create a new context for such requests.
So we introduced DefaultManager (a singleton) to get access to a manager without having a proper context around.

But we could go further... Do we really need to inject a token manager to context at all? Why not just use the default context always? It's supposed to be the same instance anyway and it's just matter of how we access it. Via injection to the context by our middleware or via a singleton - DefaultManager.

Arguments to keep both:
- We wildly use extracting the manager from the context in many places in existing code.
  Counterargument: The existing code could keep using their own inject/extract mechanism and token context but instead of using New() they could use DefaultManager from common. New services could use the new approach only.
- It's easier to test if the token manager is injected into context. Then you can inject your own test token manager into the context for particular test. It's more tricky to replace the token context in tests if it's coming from a singleton (but it's possible)

WDYT?

Reply:
> thanks for the feedback, @alexeykazakov ! I'll take your suggestions into account. For now, I do not plan to remove the logic that sets/gets the Token Manager into/from the context, but I'll add the `DefaultManager` as you suggested, as well as the `test/token`, too. I still have a couple of tests to fix, too. And for sure, I'll run the `make format-go-code` :)

actually, no: the `test/token/token.go` has some dependencies on `fabricu8-auth/account/repository` which we don't want to pull into `fabric8-common`. Could you tell me which specific functions to move to `fabric8-common` you had in mind ?

Reply:
@xcoulon everything which doesn't require account.Identity.
`EmbedTokenInContext(sub, username string)`
`GenerateToken(identityID string, identityUsername string)`
`GenerateTokenWithClaims(claims map[string]interface{})`
etc.
We also can skip functions with generate refresh tokens.